crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -310,152 +205,225 @@ impl IndexAllocator {
+        // the `warm` linked list.
+        self.unused_warm_slots -= 1;
+        self.warm
+            .remove(slot, &mut self.slot_state, |u| &mut u.unused);
+
+        // If this slot is affine to a module then additionally remove it from
+        // that module's affinity linked list. Note that if the module's affine
+        // list is empty then the module's entry in the map is completely
+        // removed as well.
+        let module = self.slot_state[slot.index()].unwrap_unused().affinity;
+        if let Some(module) = module {
+            let mut list = match self.module_affine.entry(module) {
+                Entry::Occupied(e) => e,
+                Entry::Vacant(_) => unreachable!(),
+            };
+            list.get_mut()
+                .remove(slot, &mut self.slot_state, |u| &mut u.affine);
+
+            if list.get_mut().head.is_none() {
+                list.remove();
+            }
         }
-        let i = self.rng.gen_range(0..self.free_list.len());
-        Some(self.free_list[i])
     }
 
-    /// Attempts to allocate a slot already affine to `id`, returning `None` if
-    /// `id` is `None` or if there are no affine slots.
-    fn pick_affine(&self, module_id: Option<CompiledModuleId>) -> Option<SlotId> {
-        let free = self.per_module.get(&module_id?)?;
-        free.last().copied()
+    fn pick_cold(&mut self) -> Option<SlotId> {
+        if self.last_cold == self.max_cold {
+            None
+        } else {
+            let ret = Some(SlotId(self.last_cold));
+            self.last_cold += 1;
+            ret
+        }
     }
+}
 
-    /// Remove a slot-index from the global free list.
-    fn remove_global_free_list_item(&mut self, index: SlotId) {
-        let free_list_index = self.slot_state[index.index()]
-            .unwrap_free()
-            .free_list_index();
-        assert_eq!(index, self.free_list.swap_remove(free_list_index.index()));
-        if free_list_index.index() < self.free_list.len() {
-            let replaced = self.free_list[free_list_index.index()];
-            self.slot_state[replaced.index()]
-                .unwrap_free_mut()
-                .update_free_list_index(free_list_index);
+impl List {
+    /// Creates a new one-element list pointing at `id`.
+    fn new(id: SlotId) -> List {
+        List {
+            head: Some(id),
+            tail: Some(id),
         }
     }
 
-    /// Remove a slot-index from a per-module free list.
-    fn remove_module_free_list_item(&mut self, module_id: CompiledModuleId, index: SlotId) {
-        debug_assert!(
-            self.per_module.contains_key(&module_id),
-            "per_module list for given module should not be empty"
-        );
+    /// Appends the `id` to this list whose links are determined by `link`.
+    fn append(
+        &mut self,
+        id: SlotId,
+        states: &mut [SlotState],
+        link: fn(&mut Unused) -> &mut Link,
```

Comment:
Tiny efficiency tweak, but should we monomorphize here rather than take a fn ptr? It's a small function and potentially hot in high-allocation scenarios...

Reply:
Personally I think we should defer to a future profiler, if necessary. My guess is that LLVM is already devirtualizing, and if not the mutex lock/unlock is likely much more expensive.

In general Rust, I personally think, doesn't rely enough on devirtualization and relies much too heavily on monomorphization, leading to numerous compile time issues.

Reply:
Yep, that's fine!